Added support for stats and ttl on ArrayCache

victorjonsson opened the pull request:
I had some long-time running scripts that was heavily using the ArrayCache. After a while I got some unexpected behaviour, and at first I couldn't figure out why. Imagine my surprise when I discovered that the ttl-feature was missing, it explained a lot.

So here you have my take on how it could be implemented. I also added support for hit ratio. Let me know what you think!

Enleur commented:
+1, have the same problem, so I had to use external storages and rework code to avoid frequent calls.

Tobion commented on lib/Doctrine/Common/Cache/ArrayCache.php:
```diff
@@ -59,8 +92,8 @@ protected function doContains($id)
-        $this->data[$id] = $data;
-
+        $ttl = $lifeTime > 0 ? time()+$lifeTime : false;
+        $this->data[$id] = array('content'=>$data, 'ttl'=>$ttl);
```
this should use numeric indexes instead of `content` and `ttl` to save memory. or use different arrays for data and ttl.

Tobion commented on lib/Doctrine/Common/Cache/ArrayCache.php:
```diff
@@ -38,11 +38,37 @@ class ArrayCache extends CacheProvider
+     */
+    private $stats = array();
+
+
```
extra line

Tobion commented on lib/Doctrine/Common/Cache/ArrayCache.php:
```diff
@@ -38,11 +38,37 @@ class ArrayCache extends CacheProvider
+    public function __construct()
+    {
+        $this->stats = array(
+            Cache::STATS_HITS   => 0,
```
no alignment

victorjonsson commented:
@Tobion I've changed to numeric indexes and removed the extra line. I'm unsure about what you mean by "no alignment", let me know what is incorrect and I'll fix it.

Enleur commented:
@victorjonsson I'm not quite sure about stats, it can contain large numbers in long running processes or daemons

Ocramius commented on lib/Doctrine/Common/Cache/ArrayCache.php:
```diff
@@ -38,11 +38,36 @@ class ArrayCache extends CacheProvider
+        if ($this->doContains($id)) {
+            $this->stats[Cache::STATS_HITS]++;
+            return $this->data[$id][0];
+        } else {
```
No need for `else`

Ocramius commented on lib/Doctrine/Common/Cache/ArrayCache.php:
```diff
@@ -51,7 +75,15 @@ protected function doFetch($id)
+            if ($ttl !== false && $ttl < time()) {
+                $this->doDelete($id);
+            } else {
+                return true;
```
Return early, avoid the `else`

Ocramius commented on lib/Doctrine/Common/Cache/ArrayCache.php:
```diff
@@ -38,11 +38,35 @@ class ArrayCache extends CacheProvider
     {
-        return $this->doContains($id) ? $this->data[$id] : false;
+        if ($this->doContains($id)) {
+            $this->stats[Cache::STATS_HITS]++;
```
Use a separate private property for hit stats, instead of storing it in a multidimensional array

Ocramius commented on lib/Doctrine/Common/Cache/ArrayCache.php:
```diff
@@ -38,11 +38,35 @@ class ArrayCache extends CacheProvider
+    public function __construct()
+    {
+        $this->stats = array(
+            Cache::STATS_HITS => 0,
```
You should count the hits

Ocramius commented on lib/Doctrine/Common/Cache/ArrayCache.php:
```diff
@@ -38,11 +38,35 @@ class ArrayCache extends CacheProvider
+    {
+        $this->stats = array(
+            Cache::STATS_HITS => 0,
+            Cache::STATS_MISSES => 0,
```
You should count the misses

Ocramius commented on lib/Doctrine/Common/Cache/ArrayCache.php:
```diff
@@ -38,11 +38,35 @@ class ArrayCache extends CacheProvider
+            $this->stats[Cache::STATS_HITS]++;
+            return $this->data[$id][0];
+        }
+        $this->stats[Cache::STATS_MISSES]++;
```
Use a separate private property for miss stats, instead of storing it in a multidimensional array

Ocramius commented:
@victorjonsson applied changes manually and merged, thanks!
